modules/ipp/scanner.go

```diff
@@ -225,80 +211,149 @@ func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
 				//Couldn't read content
-				vals = append(vals, val)
-				return vals, err
+				//vals = append(vals, val)
+				break
 			}
 			vals = append(vals, val)
 			if err := binary.Read(buf, binary.BigEndian, &tag); err != nil {
 				//Couldn't read next valueTag
-				return vals, err
+				break
 			}
 			// FIXME: Only try to read next namelength if previous valueTag wasn't end-of-attributes-tag
 			if err := binary.Read(buf, binary.BigEndian, &nameLength); err != nil {
 				//Couldn't read next nameLength
-				return vals, err
+				break
 			}
 		}
-		return vals, nil
+		valSlice = append(valSlice, vals)
 	}
-	//The attribute was not present
-	return nil, errors.New("Attribute \"" + attrString + "\" not present.")
+	return valSlice, nil
 }
 
-func (scan *scan) tryReadAttributes(body string) {
-	bodyBytes := []byte(body)
-	// Write reported CUPS version to results object
-	if scan.results.AttributeCUPSVersion == "" {
-		if cupsVersions, err := readAttributeFromBody(CupsVersion, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": CupsVersion,
-			}).Debug("Failed to read attribute.")
-		} else if len(cupsVersions) > 0 {
-			scan.results.AttributeCUPSVersion = string(cupsVersions[0])
+type Value struct {
+	Bytes []byte `json:"raw,omitempty"`
+}
+
+type Attribute struct {
+	Name string    `json:"name,omitempty"`
+	Values []Value `json:"values,omitempty"`
+	ValueTag byte  `json:"tag,omitempty"`
+}
+
+type AttributeGroup struct {
+	Attrs []*Attribute
+	Tag byte
+}
+
+func printGroup(g *AttributeGroup) {
+	fmt.Printf("0x%x\n", g.Tag)
+	for _, attrP := range g.Attrs {
+		fmt.Printf("\t0x%x: %q\n", attrP.ValueTag, attrP.Name)
+		for _, v := range attrP.Values {
+			fmt.Printf("\t\t%v\n", v)
 		}
 	}
-	// Write reported IPP versions to results object
-	if len(scan.results.AttributeIPPVersions) == 0 {
-		if ippVersions, err := readAttributeFromBody(VersionsSupported, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": VersionsSupported,
-			}).Debug("Failed to read attribute.")
-		} else {
-			for _, v := range ippVersions {
-				scan.results.AttributeIPPVersions = append(scan.results.AttributeIPPVersions, string(v))
+}
+
+// TODO: Log every error that could come out of this
+// TODO: Determine whether errors should be ignored, debug logged, fatal, etc.
+func readAllAttributes(body *[]byte) (*[]*Attribute) {
+	buf := bytes.NewBuffer(*body)
+	var start struct {
+		Version int16
+		StatusCode int16
+		ReqID int32
+	}
+	binary.Read(buf, binary.BigEndian, &start)
+	var tag byte
+	binary.Read(buf, binary.BigEndian, &tag)
+	var groups []AttributeGroup
+	// Until encountering end-of-attributes-tag:
+	// Per-group loop
+	for tag != 0x03 {
+		// Create a new group
+		group := AttributeGroup{Tag: tag}
+		// Read the first tag:
+		binary.Read(buf, binary.BigEndian, &tag)
+		var lastTag byte
+		for tag < 0x00 || tag > 0x05 {
```

Comment:
Note: `byte` is a `uint8`, so it's never `< 0x00`; if you need a signed byte, you can use `var tag int8`